<a href="https://colab.research.google.com/github/AliAnusKhan/Assignement1_Flyrank/blob/main/work/notebooks/w01_research_question.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/AliAnusKhan/Assignement1_Flyrank/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why
Lane: CTR / Engagement Opportunity Scoring

I chose this Core lane because it maps to a common ranking/scoring ML problem in industry — identifying content with high visibility but low engagement, which builds transferable skills for data science and ML roles.

In [7]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 2. The question: decision, action, cost of a wrong call
Question: Which pages have high impressions but low CTR relative to their position, indicating an opportunity to improve engagement?

Decision: Should we prioritize this page for content/title optimization?

Action: Content editor rewrites title/meta for flagged high-opportunity pages.

Cost of wrong call: Wasting editor time on low-impact pages, or missing real traffic opportunities

In [8]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

In [9]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.
import pandas as pd

url = "https://raw.githubusercontent.com/AliAnusKhan/Assignement1_Flyrank/main/data/raw/content_refresh_anonymized.csv"
fly_rank = pd.read_csv(url)

print("Total rows:", len(fly_rank))
print(fly_rank.columns.tolist())
fly_rank.shape
fly_rank.head()



Total rows: 30000
['content_id', 'client_id', 'search_volume', 'competition', 'competition_level', 'cpc', 'content_type', 'main_intent', 'word_count', 'char_count', 'provider_used', 'model_used', 'impressions_90d', 'clicks_90d', 'pageviews_90d', 'sessions_90d', 'users_90d', 'engaged_sessions_90d', 'ai_sessions_90d', 'scroll_events_90d', 'days_with_impressions', 'days_with_sessions', 'impressions_last_30d', 'clicks_last_30d', 'sessions_last_30d', 'impressions_prev_30d', 'clicks_prev_30d', 'sessions_prev_30d', 'content_age_days', 'age_tier', 'age_tier_order', 'days_since_last_update', 'freshness_tier', 'word_count_tier', 'char_count_tier', 'ctr', 'avg_position', 'engagement_rate', 'scroll_rate', 'ai_traffic_pct', 'impression_tier', 'position_tier', 'trend_direction', 'trend_pct']


,content_id,client_id,search_volume,competition,competition_level,cpc,content_type,main_intent,word_count,char_count,...,char_count_tier,ctr,avg_position,engagement_rate,scroll_rate,ai_traffic_pct,impression_tier,position_tier,trend_direction,trend_pct
0,content_304f48230142,client_f369cb89fc,10.0,0.67,HIGH,2.05,keyword article,transactional,3221.0,20457.0,...,15000-25000,0.76,10.6,5.88,4.55,0.0,good,striking,down,-41.4
1,content_a1fb4e703a9e,client_4e07408562,90.0,0.01,LOW,0.05,keyword article,informational,2481.0,15562.0,...,15000-25000,0.05,20.3,0.00,10.00,0.0,good,page_3_5,down,-57.7
2,content_9aa793d4d895,client_7f2253d7e2,0.0,0.00,LOW,0.00,keyword article,informational,3515.0,23643.0,...,15000-25000,0.09,36.5,0.00,28.57,0.0,good,page_3_5,down,-60.9
3,content_331d6c4de07b,client_19581e27de,10.0,0.00,LOW,0.00,keyword article,commercial,NaN,NaN,...,NaN,0.49,6.2,1.28,3.45,0.0,good,page_1,stable,-13.8
4,content_d99b7a2d90ca,client_3fdba35f04,0.0,0.00,LOW,0.00,keyword article,informational,2803.0,17469.0,...,15000-25000,0.13,44.0,0.00,24.29,0.0,good,page_3_5,down,-34.7


In [10]:
import pandas as pd

# Key numbers for CTR / Engagement Opportunity Scoring lane

print("Total content pieces:", len(fly_rank))
print("Average CTR:", round(fly_rank['ctr'].mean(), 4))
print("Average position:", round(fly_rank['avg_position'].mean(), 2))

# Opportunity signal: high impressions but low CTR
high_impression_low_ctr = fly_rank[(fly_rank['impressions_90d'] > fly_rank['impressions_90d'].median()) &
                               (fly_rank['ctr'] < fly_rank['ctr'].median())]
print("Pages with high impressions but low CTR:", len(high_impression_low_ctr))
print("Percentage of total:", round(len(high_impression_low_ctr) / len(fly_rank) * 100, 1), "%")

Total content pieces: 30000
Average CTR: 0.5107
Average position: 16.34
Pages with high impressions but low CTR: 3309
Percentage of total: 11.0 %


The starter dataset contains 30,000 content pieces (one row = one piece of content per client). The average CTR across all content is 51.07%, with an average search ranking position of 16.34. Notably, 3,309 pages (11.0%) show above-median impressions but below-median CTR — indicating a meaningful, sizeable pool of content with untapped engagement opportunity. This confirms the lane is worth pursuing: there's enough signal and volume to build a scoring model that prioritizes these pages for optimization.

## 4. Careful words: what I can and can't claim

I can claim this analysis identifies statistical patterns and correlations — such as pages with high impressions but low CTR — that suggest possible optimization opportunities. I cannot claim causation: a low CTR may be driven by factors outside this dataset, such as SERP feature competition, seasonality, or mismatched search intent. This analysis supports human decision-making; it does not guarantee that fixing a flagged page will increase traffic.

In [11]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.

Markdown cell mein paste karain:

 Named a specific lane (CTR / Engagement Opportunity Scoring) and justified the choice
 Defined a clear decision, action, and cost of a wrong call
 Showed at least 2 real numbers from the starter dataset (30,000 rows, average CTR of 51.07%, 11.0% of pages showing high-impression/low-CTR opportunity)
 Explained why this needs data/ML rather than intuition alone
 Used careful, non-overclaiming language